File: 0036-Valid-Sudoku/0036-Valid-Sudoku.c
#include <stdbool.h> // for bool, true, false
#include <stdlib.h>  // for malloc

bool isValidSudoku(char** board, int boardSize, int* boardColSize)
{
    // Arrays to track digits seen in rows, columns, and boxes
    int rows[9][9] = {0};    // rows[i][digit] = 1 if seen
    int cols[9][9] = {0};    // cols[j][digit] = 1 if seen
    int boxes[9][9] = {0};   // boxes[box_index][digit] = 1 if seen

    // Loop through all cells
    for (int i = 0; i < 9; i++)          // row index
    {
        for (int j = 0; j < 9; j++)      // column index
        {
            char cell = board[i][j];    // current cell

            if (cell == '.')            // ignore empty cells
                continue;

            int num = cell - '1';       // convert '1'-'9' to 0-8 index

            // Calculate which 3x3 box this cell belongs to
            int box_index = (i / 3) * 3 + (j / 3);

            // Check if number already seen in row, column, or box
            if (rows[i][num] || cols[j][num] || boxes[box_index][num])
                return false;

            // Mark the number as seen
            rows[i][num] = 1;
            cols[j][num] = 1;
            boxes[box_index][num] = 1;
        }
    }

    return true; // No duplicates found
}

File: 0036-Valid-Sudoku/0036-Valid-Sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """

        rows = [set() for _ in range(9)]      # Track numbers in each row
        cols = [set() for _ in range(9)]      # Track numbers in each column
        boxes = [set() for _ in range(9)]     # Track numbers in each 3x3 box

        for i in range(9):                    # Loop over each row
            for j in range(9):                # Loop over each column
                val = board[i][j]             # Get the value at that cell

                if val == '.':                # If the cell is empty, skip
                    continue

                box_index = (i // 3) * 3 + (j // 3)  # Box number (0 to 8)

                if (val in rows[i] or         # Already seen in current row?
                    val in cols[j] or         # Already seen in current col?
                    val in boxes[box_index]): # Already seen in current box?
                    return False              # Invalid if duplicate

                rows[i].add(val)              # Mark number as seen in row
                cols[j].add(val)              # Mark number as seen in column
                boxes[box_index].add(val)     # Mark number as seen in box

        return True                           # If all good, it's valid

